### PR TITLE
IOI Crocodile Editorial

### PR DESCRIPTION
**Pull request by TheGamingMousse**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

i finally remembered to link the official solution lol

### CODE REVIEW

File: solutions/gold/ioi-11-crocodile.mdx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+				dist[nxt][1] = dist[nxt][0];
+				dist[nxt][0] = time + weight;
+			} else if (time + weight < dist[nxt][1]) {
+				dist[nxt][1] = time + weight;
+				pq.push({-dist[nxt][1], nxt});
+			}
+		}
+	}
+	return dist[0][1];
+}
```
Comment (freakin23) [inline solutions/gold/ioi-11-crocodile.mdx:88]: No main function?

File: solutions/gold/ioi-11-crocodile.mdx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+<Spoiler title="Hint 1">
+
+Instead of trying to traverse from our starting node to an endpoint, consider
+starting from our endpoints and try to reach our starting node.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Obviously, the crocodile will try to prevent us from taking the best path from
```
Comment (SansPapyrus683) [inline solutions/gold/ioi-11-crocodile.mdx:19]: no obviously, even if you and i both think it is

File: solutions/gold/ioi-11-crocodile.mdx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+using namespace std;
+
+int travel_plan(int n, int m, int r[][2], int l[], int k, int p[]) {
+	vector<vector<array<int, 2>>> adj(n);
+	for (int i = 0; i < m; i++) {
+		adj[r[i][0]].push_back({r[i][1], l[i]});
+		adj[r[i][1]].push_back({r[i][0], l[i]});
+	}
+
+	// we use 1e9 instead of INT_MAX to prevent overflow
```
Comment (SansPapyrus683) [inline solutions/gold/ioi-11-crocodile.mdx:62]: how cumbersome is checking for int max

File: solutions/gold/ioi-11-crocodile.mdx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		pq.push({0, p[i]});
+		dist[p[i]] = {0, 0};
+	}
+
+	while (!pq.empty()) {
+		auto [time, at] = pq.top();
+		pq.pop();
+		time *= -1;  // undoing the negative number trick
+		if (dist[at][1] < time) { continue; }
+		for (const auto [nxt, weight] : adj[at]) {
```
Comment (SansPapyrus683) [inline solutions/gold/ioi-11-crocodile.mdx:76]: you can do unpacking with arrays????

File: solutions/gold/ioi-11-crocodile.mdx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+int travel_plan(int n, int m, int r[][2], int l[], int k, int p[]) {
+	vector<vector<array<int, 2>>> adj(n);
+	for (int i = 0; i < m; i++) {
+		adj[r[i][0]].push_back({r[i][1], l[i]});
+		adj[r[i][1]].push_back({r[i][0], l[i]});
+	}
+
+	// we use 1e9 instead of INT_MAX to prevent overflow
+	const array<int, 2> DEF = {(int)1e9, (int)1e9};
```
Comment (SansPapyrus683) [inline solutions/gold/ioi-11-crocodile.mdx:63]: `DEF` is for...

File: solutions/gold/ioi-11-crocodile.mdx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	vector<array<int, 2>> dist(n, DEF);
```
Comment (SansPapyrus683) [inline solutions/gold/ioi-11-crocodile.mdx:64]: add a comment "store second best distance asw"

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
<div class="tenor-gif-embed" data-postid="24533171" data-share-method="host" data-aspect-ratio="1.77778" data-width="100%"><a href="https://tenor.com/view/fire-writing-gif-24533171">Fire Writing GIF</a>from <a href="https://tenor.com/search/fire-gifs">Fire GIFs</a></div> <script type="text/javascript" async src="https://tenor.com/embed.js"></script>

File: solutions/gold/ioi-11-crocodile.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx:19]: oops

File: solutions/gold/ioi-11-crocodile.mdx (hunk 4 above)
Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx:76]: yeah u can do structured bindings on `std::array` lol

File: solutions/gold/ioi-11-crocodile.mdx (hunk 5 above)
Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx:63]: `DEFAULT` might be a bit more appropriate, or I could just write this directly into the declaration tbh

File: solutions/gold/ioi-11-crocodile.mdx (hunk 6 above)
Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx:64]: sure

File: solutions/gold/ioi-11-crocodile.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/ioi-11-crocodile.mdx]: yeah just default
also is `(int)` necessary

Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx]: lemme double check it, i think my compiler got mad when i didnt cast it

Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx]: image
yep for some reason it is not happy with that

Comment (SansPapyrus683) [reply solutions/gold/ioi-11-crocodile.mdx]: icic

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/gold/ioi-11-crocodile.mdx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	priority_queue<array<int, 2>> pq;
+	for (int i = 0; i < k; i++) {
+		pq.push({0, p[i]});
+		dist[p[i]] = {0, 0};
+	}
+
+	while (!pq.empty()) {
+		auto [time, at] = pq.top();
+		pq.pop();
+		time *= -1;  // undoing the negative number trick
```
Comment (ryanchou-dev) [inline solutions/gold/ioi-11-crocodile.mdx:75]: j curious is the negative trick substantially faster than `greater<>`?

**Review by ryanchou-dev — APPROVED** [review]
good stuff justin 🔥

File: solutions/gold/ioi-11-crocodile.mdx (hunk 7 above)
Comment (TheGamingMousse) [reply solutions/gold/ioi-11-crocodile.mdx:75]: nah it’s just shorter to write lol

**Pull request merged**